### PR TITLE
Feature/CSL-749 Google Single Sign On

### PR DESCRIPTION
**Pull request by klown**
Code changes to allow single sign on using a user's Google credentials/account. This is still a work in progress:

- [x] remove a number of django-allauth endpoints as superfluous to Clusive, e.g., their "Sign-up" endpoint (`/accounts/signup`)
- [x] adjust requirements.txt to include django-allauth
- [x] invoke the "kind of account" (`sign_up_role`) and  "how old are you" (`sign_up_age`) forms similar to their use in the sign-up workflow.
- [x] handle user declaring their age as under 18: delete all records associated with the SSO account
- [x] handle "Cancel" in the "kind of account" and "how old are you" forms such that the user is logged out if cancelled.
  - [x] log user out for both of the above.
- [x] remove "Change password" menu item when logged in via SSO (CSL-858).
- [ ] Add collection of display name, education level, and possibly username to SSO workflow.
  - [x] Modify check for "same" email to allow it in the case of SSO.
  - [x] Modify email confirmation to not happen in the case of SSO.
  - [x] Modify check for "same" name and to allow it in the case of SSO.

@bgoldowsky @waharnum While I've marked this as a draft pull request (see list above), there is enough here to start reviewing it.

- [x] CSL-749: https://castudl.atlassian.net/browse/CSL-749
- [x] CSL-829: https://castudl.atlassian.net/browse/CSL-829
- [x] CSL-858: https://castudl.atlassian.net/browse/CSL-858

### CODE REVIEW

**Comment by klown** [comment]
To Do:

- If the user declares their age as under 18, they see the "you need to ask your parent ..." page, but if they click the "Back" button on that page, they are logged in.  Nonetheless, their role and permissions remain "Guest".  Is this the correct behaviour?  Or, should they be logged out and end up back on the home page?
- If using the normal "Create an account" workflow, no password is stored in the user record at the end of the process.  I'm still debugging that, and can't tell if it's due to my changes or something else.
- As I noted in the Jira last week, after logging in using SSO, there is a "change password" menu item associated with the user drop down menu. That is meaningless in an SSO scenario and should be at least greyed out or perhaps removed.
- Fix conflicts with the development branch

File: src/eventlog/models.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 # A user session
 class LoginSession(models.Model):
     id = models.CharField(primary_key=True, default=uuid4, max_length=36)
-    user = models.ForeignKey(to=ClusiveUser, on_delete=models.PROTECT)
+    user = models.ForeignKey(to=ClusiveUser, on_delete=models.SET_NULL, null=True)
```
Comment (bgoldowsky) [inline src/eventlog/models.py:22]: Do you have un-checked-in migration files for these model changes?
Clean build is still giving me the not-null constraint.

**Review by bgoldowsky — COMMENTED** [review]
The registration form after logging in with google still asks me to create a username and password. That seems redundant, yes? Or is it necessary for some reason?

It also asks for email, and if I put in the email of my google account, throws an error since it exists.

File: src/clusive_project/settings.py [hunk 2]
```diff
@@ -39,8 +39,49 @@
...
+            # for automatically refresh access token, see:
+            # https://django-allauth.readthedocs.io/en/latest/providers.html?highlight=refresh%20token#django-configuration
+            'access_type': 'offline'
+        }
+    }
+}
+
+# The `allauth` redirect URI from OAuth2 server back to Clusive after a
+# successful authorization is set to '/accounts/profile/', passing back the
+# access token.  But, according to the FAQ, that will result in a 404 because
```
[the remark this answers is not in the document]
Comment (klown) [reply src/clusive_project/settings.py:73]: Yes, the is an overly long, poorly worded comment.  What it is saying at this point is that `allauth` defaults to `/accounts/profile`, but does not actually handle that endpoint.  I've shortened the comment to try and make that clear, and explained the use of `LOGIN_REDIRECT_URL` and why it is set to `/account/finish_login`.

File: src/roster/templates/roster/login.html [hunk 3]
```diff
@@ -90,7 +90,28 @@ <h2 class="mb-1">Try Clusive as a Guest</h2>
...
+                    -->
+                    {% if provider.id == "openid" %}
+                        {% for brand in provider.get_brands %}
+                            <form method="get" action="{% provider_login_url provider.id openid=brand.openid_url process=process %}">
+                                <button type="submit" class="socialaccount_provider {{provider.id}} {{brand.id}}">Sign in with {{brand.name}} SSO</button>
+                            </form>
+                        {% endfor %}
+                    {% endif %}
+                    <form method="get" action="{% provider_login_url provider.id process=process scope=scope auth_params=auth_params %}">
+                        <button type="submit" class="socialaccount_provider btn btn-primary">Sign in with {{provider.name}}</button>
```
[the remark this answers is not in the document]
Comment (klown) [reply src/roster/templates/roster/login.html:109]: Right, I somehow missed that.  I've made an attempt to incorporate the styles and layout from the mockup.  There is still one issue, though.

The `allauth` technique is to read all of the `SocialApplication` records from the database, and, for each one, configure the UI for that provider, based on the information from the database record.  That almost works.  The one thing that doesn't is the `login-google` class applied to the link.  Obviously, that applies only to the "Sign in with Google" link, and is not coming from the database.

Perhaps the approach for now is to forgo the automated loop and add the one "Google" link, carefully configuring it based on the information in the database.

File: src/eventlog/models.py (hunk 1 above)
Comment (klown) [reply src/eventlog/models.py:22]: I remember having to make migrations after those model changes, but I cannot find the migration files locally.  A possibility is that this occurred around the time my laptop crashed, and I had rebuild it from backups.  It's possible that the migration file(s) were lost at that time.  Any suggestions on how to recreate them?

**Comment by klown** [comment]
> The registration form after logging in with google still asks me to create a username and password. That seems redundant, yes? Or is it necessary for some reason?
> 
> It also asks for email, and if I put in the email of my google account, throws an error since it exists.

This is most puzzling.  The registration form should not be reached in this workflow.  The only pages that should be encountered are the `/sign_up_role` form (teacher/parent/learner), and possibly the `/sign_up_age` and `/sign_up_ask_parent` forms.  Also, those are encountered only the first time the SSO user signs in.  Once the role has been saved -- assuming the `/sign_up_ask_parent` page is not reached -- any subsequent SSO should take the user right to the `/dashboard`.

I'll merge in the latest development code and see if I can find the problem.

**Comment by klown** [comment]
New problem:  The `src/frontend/home.html` uses links for the sign-on "buttons", whereas `allauth` uses a form with submit buttons.  On the face of it, that should not matter.  However, when links are used, the `LOGIN_REDIRECT_URL` is ignored with the result that the user skips the role and age dialogs and ends up at the dashboard with an unknown role.

The latest push shows two "Sign in with Google" buttons, where the top one is the link, and the bottom is the submit button.

**Comment by klown** [comment]
> The latest push shows the behaviour with two "Sign in with Google" buttons, where the top one is the link, and the bottom is the submit button.

Here is a screen shot:
GoogleSSOButtons

**Comment by klown** [comment]
@bgoldowsky The latest push contains the missing migration files, and, thanks to @mbrambilla's quick style fix, makes the button version for triggering Google SSO look correct.

**Comment by bgoldowsky** [comment]
I think the login sequence was misbehaving for me because I didn't have the migrations in.  Anyway it seems to be working as you describe now.

One thing to consider though - by skipping the last registration page entirely we are not asking (a) for a display name (the user's first name is used, but not sure if that is appropriate for all classroom teachers)  and (b) the grade ranges taught. I think we want to send users to that page, but remove the "username" and "password" questions, keeping the others.

Also, it appears that the "username" field is set to the first name of the user when they create account via Google? Not sure how that happens, and I'm worried that it might not be unique.

**Comment by klown** [comment]
> I think the login sequence was misbehaving for me because I didn't have the migrations in. Anyway it seems to be working as you describe now.

🎉

> One thing to consider though - by skipping the last registration page entirely we are not asking (a) for a display name (the user's first name is used, but not sure if that is appropriate for all classroom teachers) and (b) the grade ranges taught. I think we want to send users to that page, but remove the "username" and "password" questions, keeping the others.

If grade ranges and display names are needed to be collected/set, that can be added to the workflow.  I'll need to think about how to modify the `SignUpView` for the two workflows, SSO vs. sign-up.

> Also, it appears that the "username" field is set to the first name of the user when they create account via Google? Not sure how that happens, and I'm worried that it might not be unique.

That's done by `allauth` and it somehow uses the profile information that Google passes back after the user finished interacting with Google's login and consent dialogs.  You can see the information in the user's SocialAccount in the `extra_data` field.  But, I don't see any `username` in that structure, unless it's as simple as a lowercase version of the `given_name` that Google provides.  The `given_name` is what's used for Clusive's display name.

Something(s) for me to explore.

**Comment by klown** [comment]
The latest code implements ignoring the "duplicate" email address error message during registration for an SSO user, and bypasses the email confirmation step.

It also attempts to similarly ignore a duplicate username based on adding a `clean_username()` to `src/roster/forms.py` based on the advice in the [Django  documentation](https://docs.djangoproject.com/en/3.2/ref/forms/validation/#cleaning-a-specific-field-attribute).  But it doesn't work in the sense that the form still puts up the "username already exists" error message, and you cannot proceed unless you change the username.

It might be that django's `UserCreationForm` is designed to stop duplicate usernames not matter what.  There is something in the documentation about excluding certain form fields from verification and that might be an approach.  Or,   perhaps we need to base the registration form on something less restrictive than the `UserCreateForm` in the SSO case (?).

**Pull request merged**